pkg/server/endpoints/node/handler.go

```diff
@@ -523,11 +531,30 @@ func (h *Handler) validateAgentSVID(ctx context.Context, cert *x509.Certificate)
 	if node == nil {
 		return fmt.Errorf("agent %q is not attested", agentID)
 	}
-	if node.CertSerialNumber != cert.SerialNumber.String() {
-		return fmt.Errorf("agent %q SVID does not match expected serial number", agentID)
+
+	currentSerial, preparedSerial := resp.Node.CertSerialNumber, resp.Node.PreparedCertSerialNumber
+	if currentSerial != "" && cert.SerialNumber.String() == currentSerial {
+		return nil
 	}
 
-	return nil
+	// New SVID activation
+	if preparedSerial != "" && cert.SerialNumber.String() == preparedSerial {
+		fieldLog := h.c.Log.WithFields(logrus.Fields{"agent": agentID, "prepared_serial": preparedSerial})
+		fieldLog.Debug("Activating agent SVID")
+		err := h.updateAttestedNode(ctx, &datastore.UpdateAttestedNodeRequest{
```

Comment:
seems like we can now have two or more RPCs racing on this update....

Reply:
Yes, but both of them will be updating the entry with the same value (moving the "prepared" serial number to "current").

The "prepared" values will always be the same for all competing RPCs (since the next rotation won't start until all RPCs are finished).

An alternative could be to add another method to the datastore and make this within a single transaction. In that way, we can avoid those possibles double-writes.

What do you think?

Reply:
I guess the concern is that with our isolation level, since two or more RPCs might attempt to update the same row in the database at the same time that in the case there is more than one transactions opened against the same target row at a time the database will fail all transactions but the first.

I feel that the situation might be likely if there is an agent actively using the JWT RPCs concurrently during a rotation.

We have other places like this in the code base, this is by no means a unique problem...

A few solutions come to mind:
1) return a special gRPC status from the datastore when a conflict is encountered, and ignore that kind of failure here
2) add a new RPC to the datastore that does the read+write operation in a single transaction that also does read-modify-write conflict resolution
3) Ignore for now

Curious what others think on this.

Reply:
Hmm yea this is kind of tricky... I guess a temporary error condition (a few failed RPCs) is strictly better than a permanent one (the server thinks we have a newer SVID than we actually have), but it also doesn't seem ideal. I guess if I had to choose one of the three options, I might choose option three (I'd prefer to apply option 1 consistently which gives scope creep, and option 2 feels a little too purpose built)

I do wonder though if there's a more direct way to accomplish this, though. It seems to me that at the end of the day, what we really want is an ACK from the agent... and because we don't have one, we imply an ACK by observing first use of the new SVID. To do that though, we need to track state etc (as you've done here).

FetchX509SVID functionality is a little overloaded and I think that has contributed to some code cleanliness problems in the node api handler... Using it for both agent SVID rotation and "regular" SVID fetching no doubt contributes to some of that as different logic applies to both of those cases and now we have to detect what is going on so we can do The Right Thing.

At the risk of throwing a wrench into the discussion, I wonder if a tighter solution here is to alter/extend the node api so we can get an agent ACK. For instance, one option that comes to mind is to introduce a `RenewX509SVID` RPC that is used exclusively for renewing/refreshing agent SVIDs (and potentially other types in the future). The functional difference between it and the existing `FetchX509SVID` would then be that in the former, the requestor and the identity requested are always equal, and in the latter they are never equal and the requestor is simply an authorized party. I'm sure there are other solutions here too, along similar lines.

If we take an extend-the-node-api route, we could build ACK into the flow and use a stream. There are probably some additional concerns that I'm not considering here, but at least wanted to see how you all feel about the pros/cons of an alternate approach. To be clear, I am not saying we _must_ do it this way, just throwing it out there

Reply:
I agree that `FetchX509SVID` is a bit overloaded and we should create another RPC sooner or later to better separate regular renewals from agent renewals.

Regarding the ACK mechanism for rotations, I believe that some things we need to think about are:

+ How connection errors will be handled? For example: if the connection is dropped at some point during the RPC stream, the next rotation-retry should be able to authenticate.

+ How agent/server crashes will be handled?. For example: if an agent crashes after the `RenewX509SVID` RPC, but before the rotator sets the new SVID.

We need to, somehow, guarantee that  "retries" can authenticate. That is the main reason behind tracking both SVIDs  (current and prepared). I'm not sure if we can get rid off that tracking.

Reply:
Hmm yes... At first I thought we could push some of it off onto the agent, but after thinking more about it, this approach feels more palatable.

Starting in this release (0.9.0) we are providing a datastore compatibility guarantee, which ends up meaning that all datastore changes must be introduced at least one full release cycle prior to actually using them. Since we need a migration to add the prepared cert info the the attested nodes table, we'll have to split this PR up and introduce only the migration first. After that is in, we'll be clear to introduce this fix starting in 0.10.0. Sorry for the delay in review here, and the unexpected constraint.

Reply:
Sure, I'll split this PR to only include the datastore changes.